BPMN Errors + Messages

pinussilvestrus opened the pull request:
Closes #38 

This adds basic support for
* BPMN Errors
* BPMN Messages

Some notes to the reviewer:

* The zeebe specific message properties will be added on top in another PR (cf. #50). 
* The border below the two action options of the selects won't be implemented by now
* For now we have two `ElementUtil` which a bit of code duplication. Rationale: We don't know where the ZeebePropertiesProvider will live in the future. That's why I wanted to minimize the shared code for the providers.

![image](https://user-images.githubusercontent.com/9433996/120644906-26bd6c80-c478-11eb-84e7-fa72377401d2.png)

![image](https://user-images.githubusercontent.com/9433996/120645049-4c4a7600-c478-11eb-9b3c-81a50b8068ad.png)

![image](https://user-images.githubusercontent.com/9433996/120645541-d4308000-c478-11eb-9508-f4c1ae98f261.png)

andreasgeier reviewed (approved):
✅ &nbsp; LGTM

MaxTru commented:
I have not started reviewing the code in detail, but looking at the high-level I realized that the dropdown does not correctly work when I try to create multiple messages: the first two messages get created, but then the dropdown stays at `create new`
![createNewMsg](https://user-images.githubusercontent.com/42800119/120776189-3d210200-c524-11eb-9325-f390297fc3a9.gif)

MaxTru reviewed (changes requested):
I have not started reviewing the code in detail, but looking at the high-level I realized that the dropdown does not correctly work when I try to create multiple messages: the first two messages get created, but then the dropdown stays at `create new`
![createNewMsg](https://user-images.githubusercontent.com/42800119/120776189-3d210200-c524-11eb-9325-f390297fc3a9.gif)

pinussilvestrus commented:
Interesting. That's maybe because of the alphabetic ordering of the messages. I'll have a look 👍

pinussilvestrus commented:
Okay, it seems that the `select` component does not work properly when the order of the options got changed while focussed. That happens when adding a new message/error => order got alphabetized with the new option - added in between.

To keep it simple: I'd simply not sort alphabetically as long we don't have a proper custom select component. Wdyt @andreasgeier @MaxTru?

andreasgeier commented:
When we introduced lists we defined the general rule to set the focus to the new item after creation. We should do the same here, particularly if that helps us to keep the sort order (I don't want to lose it).

I would accept that we have this behavior also for delete (<none> selection) and selection of existing messages/errors.

pinussilvestrus commented:
I can imagine that this list behavior (auto focus) will be hard to implement for select entries, since the select + input entry technically don't know each other so far. I'll have to double check it.

Until then, the PR is blocked.

pinussilvestrus commented:
With https://github.com/bpmn-io/properties-panel-next/pull/44/commits/c68f59ff166c9aba5f8f95abfc3c0d702e41c14a I added the autofocus behavior to the reference selects when adding new elements. As discussed, I also updated the label to "Create new ..." to better indicate an action.

![Kapture 2021-06-04 at 14 45 53](https://user-images.githubusercontent.com/9433996/120803194-96e4f480-c543-11eb-9a42-fefe29d3e313.gif)

Since it's hard to test the focus, I didn't add a test case for this (like for list items).

That's ready for review (again) @andreasgeier @MaxTru

MaxTru commented on src/properties-panel/components/entries/Select.js:
```diff
@@ -1,3 +1,16 @@
+ * @param {Object} props
+ * @param {String} props.id
+ * @param {String} props.label
+ * @param {Array<Option>} [props.options]
```
`onChange` is missing

pinussilvestrus replied:
done.

MaxTru commented:
The likelihood decreased, but the select sometimes still does not create a new message or error for me:

![createNewMsg2](https://user-images.githubusercontent.com/42800119/120824282-960a8d80-c558-11eb-93ad-ac3a5641a9d0.gif)

pinussilvestrus commented:
> The likelihood decreased, but the select sometimes still does not create a new message or error for me:
> 
> ![createNewMsg2](https://user-images.githubusercontent.com/42800119/120824282-960a8d80-c558-11eb-93ad-ac3a5641a9d0.gif)

What a nasty bug, I wasn't able to reproduce by adding 30+ messages. Maybe the `preact` select component itself comes with some hickups. 

I will maybe add a stress test and see whether there is some way to reproduce it in a reliable way. I wouldn't block this PR if we see this bug recurring, but handle it in a separate issue.

pinussilvestrus commented:
@MaxTru  By the way now while seeing your gif, the " Create new" option still has the old label. In the current state, it is "Create new ...". Also the autofocus behavior is not visible.

Maybe you didn't update the branch?

MaxTru reviewed (approved):
:clap: nice
